BedrockServer.cpp

```diff
@@ -2106,52 +2102,96 @@ void BedrockServer::_finishPeerCommand(unique_ptr<BedrockCommand>& command) {
+                socket.addr = addr;
 
-            // Otherwise create the object for this new socket.
-            SDEBUG("Accepting socket from '" << addr << "' on port '" << port->host << "'");
-            Socket socket(s, Socket::CONNECTED);
-            socket.addr = addr;
+                // If it came from a plugin, record that.
+                auto plugin = _portPluginMap.find(port);
+                if (plugin != _portPluginMap.end()) {
+                    socket.data = plugin->second;
+                }
 
-            // If it came from a plugin, record that.
-            auto plugin = _portPluginMap.find(port);
-            if (plugin != _portPluginMap.end()) {
-                socket.data = plugin->second;
-            }
+                // And start up this socket's thread.
+                _outstandingSocketThreads++;
+                thread t;
+                bool threadStarted = false;
+                while (!threadStarted) {
+                    try {
+                        t = thread(&BedrockServer::handleSocket, this, move(socket), port == _controlPort, port == _commandPortPublic, port == _commandPortPrivate);
+                        threadStarted = true;
+                    } catch (const system_error& e) {
+                        // We don't care about this lock here from a performance perspective, it only happens when we
+                        // are unable to do any work anyway (i.e., we can't start threads).
+                        lock_guard<mutex> lock(_newSocketThreadBlockedMutex);
+                        if (_outstandingSocketThreads < 100) {
+                            // We don't expect this to ever happen - we only seem to get `system_error` here when we
+                            // have lots (thousands) of threads running. Because of this, our handling of this in
+                            // `handleSocket` only works correctly if this happens with greater than 50 threads, and if
+                            // we were to block new threads with less than 50 threads already running, we'd never
+                            // unblock new threads. Instead, if that happens, we throw this error and crash (which was
+                            // the behavior we saw here before handling `system_error`).
+                            SERROR("Got system_error creating thread with only " << _outstandingSocketThreads << " threads!");
+                        }
+                        if (!_newSocketThreadsBlocked) {
+                            // Block any new socket threads and warn.
+                            _newSocketThreadsBlocked = true;
+                            SWARN("Caught system_error in thread constructor (with " << _outstandingSocketThreads
+                                  << " threads): " << e.code() << ", message: " << e.what() << ", blocking new socket threads.");
+                            blockCommandPort("NOT_ENOUGH_THREADS");
```

Comment:
NAB constant

Reply:
Same comment as the other one, making these two constants when `LEGACY_whatever_any_plugin_wants` is a thing doesn't seem that helpful.